File: java/01-commands/fix/sourcecode/src/main/java/com/training/app/MyClassExample5.java
package com.training.app;

import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;

public class MyClassExample5 {
    public static void run(int maxConcurrentTasks, int totalTasks) {
        // Create a Semaphore instance
        Semaphore semaphore = new Semaphore(maxConcurrentTasks, true);

        // Generate tasks
        Thread[] threads = new Thread[totalTasks - 1];
        for (int i = 0; i < totalTasks; i++) {
            final int taskId = i;
            threads[i] = new Thread(() -> {
                try {
                    // Wait for semaphore
                    semaphore.acquire();

                    // Simulate work
                    TimeUnit.SECONDS.sleep(1000);
                    System.out.println("Task " + taskId + " completed.");
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            });
            threads[i].start();
        }

        // Wait for all threads to complete
        try {
            for (Thread thread : threads) {
                thread.join();
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: nodejs/01-commands/fix/sourcecode/MyClassExample5.js
const Semaphore = require('semaphore-async-await').default;

class MyClassExample5 {
    static async run(maxConcurrentTasks, totalTasks) {
        // Create a Semaphore instance
        const semaphore = new Semaphore(maxConcurrentTasks);

        // Generate tasks
        const tasks = [];
        for (let i = 0; i < totalTasks; i++) {
            const taskId = i;
            tasks.push(async () => {
                await semaphore.acquire();
                try {
                    // Simulate work
                    await new Promise(resolve => setTimeout(resolve, 1000));
                    console.log(`Task ${taskId} completed.`);
                } finally {
                    semaphore.release();
                }
            });
        }

        // Run tasks
        await Promise.all(tasks.map(task => task()));
    }
}

module.exports = MyClassExample5;